src/provider/provider_coarse.c

```diff
@@ -1517,6 +1517,179 @@ coarse_memory_provider_get_stats(void *provider,
     return UMF_RESULT_SUCCESS;
 }
 
+static umf_result_t coarse_memory_provider_allocation_split(void *provider,
+                                                            void *ptr,
+                                                            size_t totalSize,
+                                                            size_t firstSize) {
+    if (provider == NULL || ptr == NULL || (firstSize >= totalSize) ||
+        firstSize == 0) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_result_t umf_result;
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)provider;
+
+    if (utils_mutex_lock(&coarse_provider->lock) != 0) {
+        LOG_ERR("lockng the lock failed");
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    assert(debug_check(coarse_provider));
+
+    ravl_node_t *node = coarse_ravl_find_node(coarse_provider->all_blocks, ptr);
+    if (node == NULL) {
+        LOG_ERR("memory block not found");
+        umf_result = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err_mutex_unlock;
+    }
+
+    block_t *block = get_node_block(node);
+
+    if (block->size != totalSize) {
+        LOG_ERR("wrong totalSize");
+        umf_result = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err_mutex_unlock;
+    }
+
+    if (!block->used) {
+        LOG_ERR("block is not allocated");
+        umf_result = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err_mutex_unlock;
+    }
+
+    block_t *new_block = coarse_ravl_add_new(coarse_provider->all_blocks,
+                                             block->data + firstSize,
+                                             block->size - firstSize, NULL);
+    if (new_block == NULL) {
+        umf_result = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_mutex_unlock;
+    }
+
+    block->size = firstSize;
+    new_block->used = true;
+
+    assert(new_block->size == (totalSize - firstSize));
+
+    assert(debug_check(coarse_provider));
+
+    if (utils_mutex_unlock(&coarse_provider->lock) != 0) {
+        LOG_ERR("unlockng the lock failed");
+        return UMF_RESULT_ERROR_UNKNOWN;
```

Comment:
NIT: you can change the flow slightly to a simpler one - initializing umf_result at the beginning and calling unlock only once

Reply:
Done

Reply:
Done in the next function too.